### ISSUE TITLE
[BUG] `exclude_unknowns=False` makes evaluation fail

### ISSUE THREAD

**Issue by vlainic** [PreferredAI/cornac#444]
Hello guys, great work, way better than Microsoft! I have a small issue, please read below:

### Description
If I set `exclude_unknowns=False` then all the ratings/rankings fail with a similar error:
e.g. `IndexError('index 1840 is out of bounds for axis 0 with size 1840')`

### In which platform does it happen?
Linux 5.11.0-36-generic

### How do we replicate the issue?
This is easy, just make sure to have non-zeros for:
`Number of unknown users = 5`
`Number of unknown items = 4`
after when doing `RatioSplit`

### Expected behavior (i.e. solution)
I would still expect to get the scores even when there are non-matching indices between train_set and test_set

**Comment by tqtg** [comment]
Could you provide the full code causing the bug? We need to see which model and dataset to be able to replicate the issue.

**Comment by vlainic** [comment]
Of course, here is modified PCRL example to replicate the issue:

```
from cornac.data import GraphModality
from cornac.eval_methods import RatioSplit
from cornac.experiment import Experiment
from cornac import metrics
from cornac.models import PCRL
from cornac.datasets import amazon_office as office


# PCRL jointly models user preferences and performs deep item features learning from auxiliary data (e.g., item relations).
# The necessary data can be loaded as follows
ratings = office.load_feedback()
contexts = office.load_graph()

# Instantiate a GraphModality, it makes it convenient to work with graph (network) auxiliary information
# For more details, please refer to the tutorial on how to work with auxiliary data
item_graph_modality = GraphModality(data=contexts)

# Define an evaluation method to split feedback into train and test sets
ratio_split = RatioSplit(
    data=ratings,
    test_size=0.5,
    rating_threshold=3.5,
    exclude_unknowns=False,
    verbose=True,
    item_graph=item_graph_modality,
)

# Instantiate PCRL
pcrl = PCRL(k=100, z_dims=[300], max_iter=10, learning_rate=0.001)

# Evaluation metrics
nDgc = metrics.NDCG(k=-1)
rec = metrics.Recall(k=20)
pre = metrics.Precision(k=20)

# Put everything together into an experiment and run it
Experiment(eval_method=ratio_split, models=[pcrl], metrics=[nDgc, rec, pre]).run()
```

which gave me following non-zero unknowns:
```
Number of unknown users = 34
Number of unknown items = 907
```

**Comment by tqtg** [comment]
@saghiles would it be possible for the PCRL model to handle unknown users and items?

**Comment by saghiles** [comment]
Hi thanks for reporting this issue. The current PCRL implementation does not handle unknown users or items. The model can in principle handle that, but the inference for unseen users/items need to be derived and implemented. Hence, we have to set `exclude_unknowns=True` when working with PCRL.

**Comment by vlainic** [comment]
I see, thank you for the quick replays.

Is this written anywhere which models need `exclude_unknowns=True` or does there have to be pre-knowledge on how models work? As I believe that all models in some way could handle `exclude_unknowns=False`.

**Comment by tqtg** [comment]
Ideally, all models should be able to handle unknown users and items. However, some specific implementations didn't take that into account. We don't have such a list of models that can handle unknown cases, though the majority can do so.

**Comment by vlainic** [comment]
Yeah, that would be ideal :)
Thank you once more, I will close the issue now.
